### PR TITLE
operator: fix upgradeImages

### PR DESCRIPTION
**Pull request by hj-johannes-lee**
fixes: #1873 
Operatorhub bundle can have sha256 image tags that are put through env vars. When operator controller manager gets upgraded, its operands (plugin daemonsets) should be updated to the image in the env vars. But it has not been working properly because of wrong parsing.

Fix it to parse the image names that have sha256 tags correctly so env vars in operator can be used as intended.

### CODE REVIEW

**Comment by mythi** [comment]
> For sha256 image tags, it has not been working properly.

Can you write how you expect it to work and how it works now. IIRC we decided SHA pinning is only for the env var approach.

**Comment by hj-johannes-lee** [comment]
If image is `intel-qat-plugin@sha256:69858421d6abd38867f61844db2c3fa7275bf6c089efd547e7de98dfc0c90a1a`

Before:
`name` was `intel-qat-plugin@sha256`
and upgradeImage was not working even though envVar is properly set.

For example, let's say there is 0.30.0 version of deployed operatorhub bundle using olm (as written in the [link](https://github.com/intel/intel-device-plugins-for-kubernetes/blob/main/DEVEL.md#publish-a-new-version-of-the-intel-device-plugins-operator-to-operatorhubio)), and then let's say we deploy a qat plugin (0.30.0 version).
And then we update the bundle with the comamnd:
```
operator-sdk run bundle-upgrade intel/intel-deviceplugins-operator:0.31.0
```
Then, the qat-plugin is expected to be upgraded automatically according to the envVar, but it does not because the parsing has been wrong (`intel-qat-plugin@sha256`)

So,
I expect `name` to be `intel-qat-plugin` so that it can be updated when another sha256 is found from envVar.
And it works now like that, and tested with olm.

**Comment by mythi** [comment]
Thanks for noticing this! TestUpgrade unit test needs updating to get this covered. Also .1 release is needed

**Comment by hj-johannes-lee** [comment]
Yeah, I think we should have had it early... It's my mistake.
I pushed the commit for the testUpgrade fix too.

File: pkg/controllers/reconciler_test.go [hunk 1]
```diff
@@ -61,11 +66,19 @@ func TestUpgrade(test *testing.T) {
...
+			initimage:         initimage + "@sha256:" + preSha256,
+			expectedInitimage: initimage + "@sha256:" + sha256,
+			upgrade:           true,
+		},
 	}
 
 	for i := range tests {
 		t := tests[i]
-
+		os.Setenv(strings.ReplaceAll(strings.ToUpper(filepath.Base(t.image)), "-", "_")+"_SHA", t.image+"@sha256:"+sha256)
```
[the remark this answers is not in the document]
Comment (hj-johannes-lee) [reply pkg/controllers/reconciler_test.go:80]: Umm what would be the way to put the `flag`?
another `bool` type field after `upgrade bool`?

[the remark this answers is not in the document]
Comment (hj-johannes-lee) [reply pkg/controllers/reconciler_test.go]: I put `isEnvVarSet`. If it's not what you meant, please comment again.!

[the remark this answers is not in the document]
Comment (hj-johannes-lee) [reply pkg/controllers/reconciler_test.go]: I do not understand what you mean.

[the remark this answers is not in the document]
Comment (hj-johannes-lee) [reply pkg/controllers/reconciler_test.go]: I think I understood what you mean, and changed.
But, I am not sure if it is 'simpler' though I think it looks 'clearer'.

One thing I am not sure is if it would look better if I make a function for
`strings.ReplaceAll(strings.ToUpper(filepath.Base(name)), "-", "_") + "_SHA"` part?

File: pkg/controllers/reconciler_test.go [hunk 2]
```diff
@@ -61,11 +67,24 @@ func TestUpgrade(test *testing.T) {
...
+			upgrade:           true,
+		},
 	}
 
 	for i := range tests {
 		t := tests[i]
 
+		for key, value := range t.envVars {
+			os.Setenv(key, value)
+		}
```
Comment (tkatila) [inline pkg/controllers/reconciler_test.go:86]: Env variables stay from test run to test run. I think we should somehow clear the variables after the run. If more tests are added which do not use the env variables, they could get false-negative results.

Comment (hj-johannes-lee) [reply pkg/controllers/reconciler_test.go:86]: Thanks! Added `os.clearEnv()`!

File: pkg/controllers/reconciler_test.go [hunk 3]
```diff
@@ -61,18 +67,33 @@ func TestUpgrade(test *testing.T) {
...
+
 		upgrade := UpgradeImages(context.Background(), &t.image, &t.initimage)
 
 		if !(upgrade == t.upgrade && t.image == t.expectedImage && t.initimage == t.expectedInitimage) {
 			test.Errorf("expectedUpgrade: %v, received: %v", t.upgrade, upgrade)
 			test.Errorf("expectedImage: %s, received: %s", t.expectedImage, t.image)
 			test.Errorf("expectedInitimage: %s, received: %s", t.expectedInitimage, t.initimage)
 		}
+
+		os.Clearenv()
```
Comment (tkatila) [inline pkg/controllers/reconciler_test.go:96]: This will delete everything, it should only remove the ones that were added.

Comment (hj-johannes-lee) [reply pkg/controllers/reconciler_test.go]: fixed to `os.UnsetEnv`

File: pkg/controllers/reconciler_test.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (tkatila) [reply pkg/controllers/reconciler_test.go]: I think having a function for the ReplaceAll-ToUpper combo would make it look better. Especially with the test code. Feel free to make the change.

File: pkg/controllers/reconciler_test.go [hunk 4]
```diff
@@ -61,18 +67,35 @@ func TestUpgrade(test *testing.T) {
 			expectedInitimage: initimage,
 			upgrade:           false,
 		},
+		{
+			envVars: map[string]string{strings.ReplaceAll(strings.ToUpper(filepath.Base(image)), "-", "_") + "_SHA": image + "@sha256:" + sha256,
+				strings.ReplaceAll(strings.ToUpper(filepath.Base(initimage)), "-", "_") + "_SHA": initimage + "@sha256:" + sha256},
```
Comment (mythi) [inline pkg/controllers/reconciler_test.go:72]: This is rather complex for the reader to understand what this test is all about. I'd suggest to spell out the strings for readability, maybe with a different image name:
```suggestion
			envVars: map[string]string{
			        INTEL_QAT_PLUGIN_SHA: "intel/intel-qat-plugin@sha256:000000000000000000000000000000000000000000000000000000000000000a",
			        INTEL_QAT_INITIMAGE_SHA: "intel/intel-qat-initimage@sha256:000000000000000000000000000000000000000000000000000000000000000a",
				},
```

**Review by mythi — CHANGES_REQUESTED** [review]
for full coverage, we'd also need a case where `image` is same as the corresponding envVar and when the envVar is not in the pinned/sha256 format.

File: pkg/controllers/reconciler_test.go (hunk 4 above)
Comment (hj-johannes-lee) [reply pkg/controllers/reconciler_test.go:72]: If you see the image name, it is fixed as dsa plugin's image name. Is the qat example just an example?

**Comment by hj-johannes-lee** [comment]
@mythi 
> for full coverage, we'd also need a case where image is same as the corresponding envVar and when the envVar is not in the pinned/sha256 format.

Umm, we do not check in `reconciler.go` whether the envVar value is sha256 format or not.
In our current implementation, we always expect the envvar value for the image would be in sha256 format.
The name of the envvar itself is like `INTEL_DSA_PLUGIN_SHA`

So, do you want to improve the implementation? Or... did I misunderstand what you mean?

File: pkg/controllers/reconciler_test.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (hj-johannes-lee) [reply pkg/controllers/reconciler_test.go]: @tkatila Thanks for your comment.!

I was thinking that looks better, but Mikko suggested to "hardcode" that part.
I thought it makes sense and looks intuitive to readers. So, let me just.. use that.

**Comment by mythi** [comment]
> So, do you want to improve the implementation?

we can focus on the current functionality and make sure all scenarios are covered

**Comment by hj-johannes-lee** [comment]
@mythi
Thanks for the comment. I think we all cover for the "current implementation."

**Comment by hj-johannes-lee** [comment]
Fixed.!

And,, added some comments in the code to make it understandable more easily (actually for myself to understand next time.. I need to always follow the process and understand again...). If you think it's not good, let me konw.!

**Review by mythi — APPROVED** [review]
LGTM, thanks!

**Review by tkatila — APPROVED** [review]
